flocker/testtools/_base.py

```diff
@@ -104,6 +111,153 @@ def mktemp(self):
+            logs are stored.
+        :param detail_name: Name of the Twisted log detail.
+        """
+        twisted_log = detailed.getDetails()[detail_name]
+        split_logs = [None]
+
+        def _get_split_logs():
+            # Memoize the split log so we don't iterate through it twice.
+            if split_logs[0] is None:
+                split_logs[0] = _split_map_maybe(
+                    _get_eliot_data, _iter_content_lines(twisted_log),
+                )
+            return split_logs[0]
+
+        # The trick here is that we can't iterate over the base detail yet.
+        # We can only use it inside the iter_bytes of the Content objects
+        # that we add. This is because the only time that we *know* the
+        # details are populated is when the details are evaluated. If we call
+        # it in _setUp(), the logs are empty. If we call it in cleanup, the
+        # detail is gone.
+
+        detailed.addDetail(
+            detail_name,
+            Content(UTF8_TEXT, lambda: _get_split_logs()[0]))
+
+        detailed.addDetail(
+            self._ELIOT_LOG_DETAIL_NAME,
+            Content(
+                UTF8_TEXT, lambda: _prettyformat_lines(_get_split_logs()[1])))
+
+
+def _split_map_maybe(function, sequence, marker=None):
+    """
+    Lazily map ``function`` over ``sequence``, yielding two streams:
+    ``(original, applied)``
+
+    :param function: Unary callable that might return ``marker``.
+    :param sequence: Iterable of objects that ``function`` will be applied to.
+    :param marker: Value returned by ``function`` when it cannot be
+        meaningfully applied to an object in ``sequence``.
+    :return: ``(original, applied)``, where ``original`` is an iterable of all
+        the elements, ``x``, in ``sequence`` where ``function(x)`` is
+        ``marker``, and ``applied`` is an iterable of all of the results of
+        ``function(x)`` that are not ``marker``.
+    """
+    annotated = ((x, function(x)) for x in sequence)
+    original, mapped = tee(annotated)
+    return (
+        (x for (x, y) in original if y is marker),
+        (y for (x, y) in mapped if y is not marker)
+    )
+
+
+def _prettyformat_lines(lines):
+    """
+    Pretty format lines of Eliot logs.
+    """
+    for line in lines:
+        data = json.loads(line)
+        yield pretty_format(data) + '\n'
+
+
+_ELIOT_MARKER = ' [-] ELIOT: '
+_ELIOT_MARKER_LENGTH = len(_ELIOT_MARKER)
+
+
+def _get_eliot_data(twisted_log_line):
+    """
+    Given a line from a Twisted log message, return the text of the Eliot log
+    message that is on that line.
+
+    If there is no Eliot message on that line, return ``None``.
+
+    :return: A logged eliot message without Twisted logging preamble, or
+        ``None``.
+    :rtype: unicode or ``NoneType``.
+    """
+    index = twisted_log_line.find(_ELIOT_MARKER)
+    if index < 0:
+        return None
+    return twisted_log_line[index + _ELIOT_MARKER_LENGTH:].strip()
```

Comment:
:question: This could be also written as
```python
_, _, eliot_data = twisted_log_line.partition(_ELIOT_MARKGER)
if eliot_data:
    return eliot_data.strip()
else:
    return None
```
I'm not sure which is clearer.